### PR TITLE
Project presentation for CR

### PR DESCRIPTION
**Pull request by sergio-usma**
# **🚀 Ruby on Rails Group Project  - :gem: Recipe App**

image

## Description 📝
The Recipe app serves as a comprehensive solution for managing recipes, ingredients, and inventory. The primary goal is to provide users with a seamless experience in saving ingredients, tracking their inventory, creating recipes, and generating shopping lists based on available ingredients and missing items. Additionally, the app prioritizes the sharing of recipes, allowing users to make them public for widespread accessibility.

### Database Schema Diagram :diamond_shape_with_a_dot_inside: 

image

## Features 🛠️
Data Model Update 🗂️

Incorporated a data model aligned with the provided Entity Relationship Diagram.
Adjusted data model complexity based on team size, ensuring seamless collaboration.
Data Model

### Added common interfaces 🖥️

- [X] Login Page 🚪
- [X] Registration Page 📝
- [X] Recipes List (with CRUD operations, excluding 'update') 📋
- [X] Public Recipe List (ordered by newest) 🌐
- [X] Design Personalization 🎨
- [X] Maintained the wireframes layout.
- [X] Personalized design elements including colors, typographies, and spacings.

## Page-specific Functionality 🔄

- [X] Login and Registration Pages:

- Implemented with Devise for secure authentication.

- [X] Recipes List:

- Displays recipes created by the logged-in user.
- Leads to recipe details.

- [X] Public Recipe List:

- Displays all public recipes ordered by newest.
- Allows the owner to delete their recipe.
- Leads to recipe details.

- [X] Recipe Details:

- Toggle button for making a recipe public or private.
- Displays recipe details if public or user is the owner.
- Leads to the form for adding new food if the user is the owner.

### Navigation Menu 🚢

Created a user-friendly navigation menu for seamless access to all app pages.

## Unit and Integration Tests 🧪

image

- [X] Comprehensive testing implemented for:
- [X] Login and registration functionality.
- [X] CRUD operations on user-specific recipes.
- [X] Proper display and functionality of public recipes.
- [X] Toggle functionality and details display in recipe details.
- [X] Food list display and form functionality.
- [X] General shopping list accuracy in item count and total price.

## Additional Considerations 🤔

- Ensured no N+1 queries for optimal performance.
- Implemented a detailed food list display, including quantity.
- General shopping list view now shows missing food items for all recipes, along with total count and price.

### Checklist ✅

- [X]  Data model alignment with provided diagrams.
- [X]  Personalized design elements.
- [X]  Proper implementation of Devise for authentication.
- [X]  Functional navigation menu.
- [X]  Comprehensive unit and integration tests.
- [X]  No N+1 queries.

<hr>
:building_construction: Made by Sergio Usma @sergio-usma & Gumaro Monroy @Movgto

### CODE REVIEW

File: app/controllers/general_shopping_lists_controller.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    required_foods = {}
+    @recipes.each do |recipe|
+      RecipeFood.includes(:recipe).where(recipe:).references(:recipe).each do |recipe_food|
+        if required_foods[recipe_food.food_id].nil?
+          required_foods[recipe_food.food_id] = recipe_food.quantity
+        else
+          required_foods[recipe_food.food_id] += recipe_food.quantity
+        end
+      end
+    end
```
Comment (Ibizugbe) [inline app/controllers/general_shopping_lists_controller.rb:36]: Nice job so far😉. currently, the generate shopping list feature isn't working as required😉. I should be able to see the list of food that is missing for all recipes of the logged-in user and see the total food items and total price of the missing food😉. Kindly do well to add this feature😉.

image

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @sergio-usma  @Movgto 👋🏼👋🏼👋🏼, 

request change


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights✅✅✅

- [x] Linters are passing👍🏼.
- [x] Unit and integration tests have been written with all test passing👍🏼.
- [x] Nice job ensuring your project matches the wireframe👍🏼.
- [x] Devise has been used for authentication👍🏼.


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/general_shopping_lists_controller.rb (hunk 1 above)
Comment (sergio-usma) [reply app/controllers/general_shopping_lists_controller.rb:36]: Please check this video https://www.loom.com/share/e6577bdef3914b4baa1d6b3dc9f8fa48?sid=9c322b27-1873-4b2a-b178-e36f84a90f84

**Review by franclobo — APPROVED** [review]
Hi @sergio-usma and @Movgto, 

# STATUS: APPROVED 😃 

git-merge

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 


## Highlights

- No linter errors ✔️ 
- Good UI ✔️
- Correct PostgreSQL database ✔️ 
- Correct authorization and authentication ✔️ 

Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**